### PR TITLE
add enum description

### PR DESCRIPTION
**Pull request by JieAnthony**

### CODE REVIEW

**Comment by stancl** [comment]
Can you provide some good use cases for this?

**Comment by JieAnthony** [comment]
@stancl [https://github.com/BenSampo/laravel-enum#static-getdescriptionmixed-value-string](https://github.com/BenSampo/laravel-enum#static-getdescriptionmixed-value-string)

For applications whose native language is not English, the enum needs to show the description

File: src/DescriptionInterface.php [hunk 1]
```diff
@@ -0,0 +1,8 @@
+<?php
+
+namespace ArchTech\Enums;
+
+interface DescriptionInterface
```
Comment (stancl) [inline src/DescriptionInterface.php]: I think an interface isn't needed. You can add an abstract method to the trait

**Comment by stancl** [comment]
I'm thinking if there should be a more general solution to mapping cases to other values. For instance, I have a `getColor()` method in a `TaskStatus` enum

**Comment by stancl** [comment]
See the failing checks

**Comment by JieAnthony** [comment]
@stancl 
fixed
I am neutral on this PR
Thank you😍

**Comment by samlev** [comment]
I feel like `getDescription()` should just be called `description()`, or possibly be accessible through `__get()` so that `$enum->description` exists

**Comment by stancl** [comment]
I agree, `description()` sounds better.

**Comment by stancl** [comment]
Defining `__get()` would be problematic though because then you couldn't add e.g. `color` if you already had `__get()` overridden by the trait for `description`. So instead, we want to find a way to **provide these key-value pairs to `__get()`**.

(Update: We need to use `__call()`, not `__get()`, since that is not allowed on enums)

As I wrote in a previous comment:
> I'm thinking if there should be a more general solution to mapping cases to other values. For instance, I have a `getColor()` method in a `TaskStatus` enum

One solution that came to mind is using attributes. I just tried it in Tinkerwell, and it seems to work pretty well.

```php
abstract class EnumMetaProperty
{
    public function __construct(
        public mixed $value,
    ) {}

    /** Get the name of the accessor method */
    public static function method(): string
    {
        $parts = explode('\\', static::class);

        return lcfirst(end($parts));
    }
}

#[Attribute(Attribute::TARGET_CLASS)]
class EnumMeta
{
    public function __construct(
        /** @var EnumMetaProperty[] */
        public array $metaProperties = [] // todo perhaps support variadic args to remove extra [] from user code
    ) {}
}

trait Metadata
{
    public function __call(string $property, $arguments): mixed
    {
        $enumReflection = new ReflectionObject($this);
        $enumMeta = array_values(array_filter(
          	$enumReflection->getAttributes(),
            fn (ReflectionAttribute $attr) => $attr->getName() === EnumMeta::class,
        ))[0]?->newInstance();

        $metaProperties = $enumMeta?->metaProperties ?? [];

        foreach ($metaProperties as $metaProperty) {
            if ($metaProperty::method() === $property) {
                return array_values(array_filter(array_map(function (string $caseName) use ($metaProperty) {
                    $caseAttributes = (new ReflectionEnumUnitCase(static::class, $caseName))->getAttributes();
                    $caseAttributes = array_map(fn (ReflectionAttribute $attr) => $attr->newInstance(), $caseAttributes);
                    $usedAttribute = array_values(array_filter($caseAttributes, fn (EnumMetaProperty $attribute) => $attribute::class === $metaProperty))[0] ?? null;
                    return $usedAttribute?->value;
                }, array_column(static::cases(), 'name'))))[0] ?? null;
            }
        }

        return null;
    }
}

// ------ userland code ------

// The user creates attributes for the individual properties like this
#[Attribute]
class Description extends EnumMetaProperty
{
    // Optional; the name falls back to snakeCase of the class name
    // public static function method(): string
    // {
    //     return 'description';
    // }
}

#[Attribute]
class Color extends EnumMetaProperty {}

#[EnumMeta([Description::class, Color::class])]
enum TaskStatus: int
{
    use Metadata;

    #[Description('Incomplete Task')] #[Color('red')]
    case INCOMPLETE = 0;

    #[Description('Completed Task')] #[Color('green')]
    case COMPLETE = 1;
}

TaskStatus::INCOMPLETE->description(); // Incomplete task
TaskStatus::INCOMPLETE->color(); // red
```

(The internals are ugly & the naming isn't perfect, this is just for demo purposes).

If this would solve the `->description()` for you, I can add this to the core.

The only issue I see is that enum parameters can't be dynamic, so if you wanted to use e.g. `__()` for localization, you'd do this:

```php
class Color extends EnumMetaProperty
{
    public function __construct(public string $value)
    {
        $this->value = __($value);
    }
}


#[EnumMeta([Description::class])]
enum TaskStatus: int
{
    use Metadata;

    #[Description('statuses.incomplete')]
    case INCOMPLETE = 0;

    #[Description('statuses.complete')]
    case COMPLETE = 1;
}
```

To have `TaskStatus::INCOMPLETE->description()` work like `__('statuses.incomplete')`.

If you wanted to use `->description()` with the `match()` syntax inside, it's best to do that without any traits I think.

Also, the solution I proposed would also allow things like:
- `fromColor()`
- `tryFromColor()`

Since we override `__get()` and we can convert cases to colors and vice versa.

**Comment by JieAnthony** [comment]
@stancl great idea

**Comment by stancl** [comment]
Closing in favor of #8 :)

**Pull request closed without merging**